chore(weave): Improve OpenTelemetry examples and add non-instrumented example (#4115)
wandb/weave: docs/docs/guides/tools/otel.md

# Send OpenTelemetry Traces to Weave

## Overview
Weave supports ingestion of OpenTelemetry compatible trace data through a dedicated endpoint. This endpoint allows you to send OTLP (OpenTelemetry Protocol) formatted trace data directly to your Weave project.

> ## Endpoint Details

Reviewer: ```suggestion
## Endpoint details
```

**Path**: `/otel/v1/traces`
**Method**: POST
**Content-Type**: `application/x-protobuf`

## Authentication

> Standard Weights & Biases authentication is used. You must have write permissions to the project where you're sending trace data.

Reviewer: ```suggestion
Standard W&B authentication is used. You must have write permissions to the project where you're sending trace data.
```

## Required Headers
- `project_id: <your_entity>/<your_project_name>`
- `Authorization=Basic <Base64 Encoding of api:$WANDB_API_KEY>`

## Examples:

In the below examples there are a few fields you must modify before the code will run:
1. WANDB_API_KEY: Obtain this from https://wandb.ai/authorize
2. Entity: You can only log traces to project under an entity you have access to. You can find your entity name by visiting https://wandb.ai/home and checking the Teams field in the left sidebar
3. Project Name: Choose a fun name!
4. OPENAI_API_KEY: Obtain this from your OPENAI dashboard

### OpenInference Instrumentation:

This example shows how to use the OpenAI instrumentation. There are many more available which you can find in the official repository: https://github.com/Arize-ai/openinference

> First install the required dependencies:

Reviewer: ```suggestion
First, install the required dependencies:
```

}

tracer_provider = trace_sdk.TracerProvider()

# Configure the OTLP exporter
exporter = OTLPSpanExporter(
    endpoint=OTEL_EXPORTER_OTLP_ENDPOINT,
    headers=OTEL_EXPORTER_OTLP_HEADERS,
)

# Add the exporter to the tracer provider
tracer_provider.add_span_processor(SimpleSpanProcessor(exporter))

# Optionally, print the spans to the console.
tracer_provider.add_span_processor(SimpleSpanProcessor(ConsoleSpanExporter()))

OpenAIInstrumentor().instrument(tracer_provider=tracer_provider)

def main():
    client = openai.OpenAI(api_key=OPENAI_API_KEY)
    response = client.chat.completions.create(
        model="gpt-3.5-turbo",
        messages=[{"role": "user", "content": "Describe OTEL in a single sentence."}],
        max_tokens=20,
        stream=True,
        stream_options={"include_usage": True},
    )
    for chunk in response:
        if chunk.choices and (content := chunk.choices[0].delta.content):
            print(content, end="")

if __name__ == "__main__":
    main()
```

Finally, once you have set the fields specified above to their correct values, run the code:

```bash
python openinference_example.py
```

### OpenLLMetry Instrumentation:

> This example shows how to use the OpenAI instrumentation. There are many more available which you can find in the official repository: https://github.com/traceloop/openllmetry/tree/main/packages

Reviewer: ```suggestion
The following example shows how to use the OpenAI instrumentation. Additional examples are available at [https://github.com/traceloop/openllmetry/tree/main/packages](https://github.com/traceloop/openllmetry/tree/main/packages).
```

OTEL_EXPORTER_OTLP_ENDPOINT = f"{WANDB_BASE_URL}/otel/v1/traces"

# Can be found at https://wandb.ai/authorize
WANDB_API_KEY = "<your-wandb-api-key>"
AUTH = base64.b64encode(f"api:{WANDB_API_KEY}".encode()).decode()

OTEL_EXPORTER_OTLP_HEADERS = {
    "Authorization": f"Basic {AUTH}",
    "project_id": PROJECT_ID,
}

tracer_provider = trace_sdk.TracerProvider()

# Configure the OTLP exporter
exporter = OTLPSpanExporter(
    endpoint=OTEL_EXPORTER_OTLP_ENDPOINT,
    headers=OTEL_EXPORTER_OTLP_HEADERS,
)

# Add the exporter to the tracer provider
tracer_provider.add_span_processor(SimpleSpanProcessor(exporter))

# Optionally, print the spans to the console.
tracer_provider.add_span_processor(SimpleSpanProcessor(ConsoleSpanExporter()))

trace.set_tracer_provider(tracer_provider)
# Creates a tracer from the global tracer provider
tracer = trace.get_tracer(__name__)
tracer.start_span('name=standard-span')

def my_function():
    with tracer.start_as_current_span("outer_span") as outer_span:
        client = openai.OpenAI()
        input_messages=[{"role": "user", "content": "Describe OTEL in a single sentence."}],
        # This will only appear in the side panel

>         outer_span.set_attribute("input.value", json.dumps(input_att))

Reviewer 2: Hey @zbirenbaum, `input_att` is not defined. Do you mean `input_messages`?

    main()
```

Finally, once you have set the fields specified above to their correct values, run the code:

```bash
python openllmetry_example.py
```

### Without Instrumentation

> If you would prefer to use the OTEL directly instead of an instrumentation package you may do so. Attributes will be parsed according to the OpenTelemetry semantic conventions: https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-spans/
> 
> First install the required dependencies:
> 
> ```bash
> pip install openai opentelemetry-sdk opentelemetry-api opentelemetry-exporter-otlp-proto-http
> ```
> 
> Next, paste the following code into a python file such as `opentelemetry_example.py`

Reviewer: ```suggestion
If you prefer to use the OTEL directly instead of an instrumentation package, you may do so. Attributes will be parsed according to the OpenTelemetry semantic conventions described at [https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-spans/](https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-spans/).

First, install the required dependencies:

```bash
pip install openai opentelemetry-sdk opentelemetry-api opentelemetry-exporter-otlp-proto-http
```

Next, paste the following code into a python file such as `opentelemetry_example.py`
```

Author: I think this one deleted some stuff it shouldn't have so I applied the suggestion manually. Let me know if it doesn't match what you were going for

Reviewer: Looked at it via the Edit button, and I think the GH UI is rendering it wrong because of a code block inside the suggestion block. So this should be ok to go still

Reviewer: This is the full MD copyed from the comment:

```suggestion
If you prefer to use the OTEL directly instead of an instrumentation package, you may do so. Attributes will be parsed according to the OpenTelemetry semantic conventions described at [https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-spans/](https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-spans/).

First, install the required dependencies:

```bash
pip install openai opentelemetry-sdk opentelemetry-api opentelemetry-exporter-otlp-proto-http
```

Next, paste the following code into a python file such as `opentelemetry_example.py`
```

Author: Since I already applied it manually it won't let me commit that anymore :( I think its the same, all that changed was the comma and the first paragraph right?

Reviewer: Ah got it, reapplied in [8c75fe3](https://github.com/wandb/weave/pull/4115/commits/8c75fe392aa56ff2414f1c45d70656a4eb660a66)

> If you prefer to use the OTEL directly instead of an instrumentation package, you may do so. Attributes will be parsed according to the OpenTelemetry semantic conventions described at [https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-spans/](https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-spans/).
> 
> First, install the required dependencies:
> 
> ```bash
> pip install openai opentelemetry-sdk opentelemetry-api opentelemetry-exporter-otlp-proto-http
> ```
> 
> Next, paste the following code into a python file such as `opentelemetry_example.py`

Reviewer: ```suggestion
If you would prefer to use OTEL directly instead of an instrumentation package, you may do so. Attributes will be parsed according to the OpenTelemetry semantic conventions described at [https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-spans/](https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-spans/).

First, install the required dependencies:

```bash
pip install openai opentelemetry-sdk opentelemetry-api opentelemetry-exporter-otlp-proto-http
```

Next, paste the following code into a python file such as `opentelemetry_example.py`
```

def my_function():
    with tracer.start_as_current_span("outer_span") as outer_span:
        client = openai.OpenAI()
        input_messages=[{"role": "user", "content": "Describe OTEL in a single sentence."}],
        # This will only appear in the side panel
        outer_span.set_attribute("input.value", json.dumps(input_att))
        # This follows conventions and will appear in the dashboard
        outer_span.set_attribute("gen_ai.system", 'openai')
        response = client.chat.completions.create(
            model="gpt-3.5-turbo",
            messages=input_messages,
            max_tokens=20,
            stream=True,
            stream_options={"include_usage": True},
        )
        out = ""
        for chunk in response:
            if chunk.choices and (content := chunk.choices[0].delta.content):
                out += content
        # This will only appear in the side panel
        outer_span.set_attribute("output.value", json.dumps({"content": out}))

if __name__ == "__main__":
    my_function()
```

Finally, once you have set the fields specified above to their correct values, run the code:

```bash
python opentelemetry_example.py
```

> The attribute prefixes `gen_ai` and `openinference` are used to determine which convention, if any, to interpret the trace as. If neither key is detected then all attributes will be visible in the trace view. The full span will always be available in the side panel when you select a trace.

Reviewer: ```suggestion
The attribute prefixes `gen_ai` and `openinference` are used to determine which convention to use, if any, when interpreting the trace. If neither key is detected, then all attributes are visible in the trace view. The full span is available in the side panel when you select a trace.
```
